### DISCUSSION TITLE
Version extraction: non-contiguous match groups

### DISCUSSION THREAD

**Discussion by foolishgrunt** [ImranR98/Obtainium#1892] [Q&A] [answered]
Consider a case such as [Collabora](https://collaboraoffice.com/downloads/Collabora-Office-Android-Snapshot): the apk filename is `collabora-office-mobile-24.04-snapshot-arm64-v8a-2024-10-07.apk`.

Since this is a (semi-)weekly build of a LTS release, a sane version number would be something along the lines of `[release number]-[build date]`, e.g. `24.04-2024-10-07`. Both of these strings are contained within the filename, but non-contiguously.

It's my understanding that there is no native RegEx provision for capturing non-contiguous text in a single match group, but can anyone think of a way to work around this in Obtainium to produce the desired version number?

**Comment by DarkCrypt** [comment]
Version String Extraction ReGex: `[0-9]+\.[0-9]+`

**Reply by foolishgrunt** [reply]
That returns a value of `24.04`: only the first of two desired strings.

**Reply by DarkCrypt** [reply]
Apologies, I somehow misunderstood the question at first. I don't think this is possible unless Obtainium could support multiple match groups.

**Reply by DwainZwerg** [reply]
1. https://collaboraoffice.com/downloads/Collabora-Office-Android-Snapshot is the snapshot version of CollaboraOffice. The stable LTS version is on the F-Droid-Third-Party https://www.collaboraoffice.com/downloads/fdroid/repo/. Perhaps I should adjust the entry in the website so that it doesn't cause any further confusion.
2. If you set the version string extraction regex to `[0-9]+\.[0-9]+-snapshot-arm64-v8a-[0-9]+-[0-9]+-[0-9]`, it should do exactly what you want.

**Reply by DarkCrypt** [reply]
@DwainZwerg 
~~Your ReGex has `snapshot` misspelled.~~

@foolishgrunt 
DwainZwerg's above ReGex produces: `24.04-snapshot-arm64-v8a-2024-10-07`.

You want: `24.04-2024-10-07`.

Is this correct or am I still misunderstanding?

**Reply by foolishgrunt** [reply]
That's correct, I'm looking to include just the version and the build date, not the intermediate string about architecture.

And this is relevant for even the stable repo, because it also contains multiple builds of the same release.

**Reply by DarkCrypt** [reply]
Ok, I did understand you the 2nd time. Unfortunately, I'm having to revert to my 2nd comment above.

@ImranR98 
I find what foolishgrunt is talking about here very interesting. I'd like to hear your input about this.

**Reply by ImranR98** [reply]
Sure, makes sense. We could support multiple match groups. In the "match groups to use" field you'd be able to enter a comma separated list of groups - like `1,2` for the first and second groups (Obtainium would add a period between them).

**Reply by foolishgrunt** [reply]
That would work for my use case.

**Reply by DarkCrypt** [reply]
Sounds great to me.

**Comment by ImranR98** [comment]
Opened #1913

**Reply by ImranR98** [reply, marked as answer]
Just realized this was already implemented in #1235.
Regex: `([0-9]+\.[0-9]+)-snapshot-arm64-v8a-([0-9]+-[0-9]+-[0-9]+)`
Match groups: `$1.$2`

**Reply by foolishgrunt** [reply]
Confirmed it works, thanks!

**Reply by DwainZwerg** [reply]
What does this config look like (please share as a link)?

**Reply by foolishgrunt** [reply]
> What does this config look like (please share as a link)?

Pulling from the stable repo:
[087-11ef-a420-61cc893eff76.json](https://github.com/user-attachments/files/17478618/087-11ef-a420-61cc893eff76.json)

**Reply by DarkCrypt** [reply]
Awesome! Thanks for the info!

Config for arm64 following ImranR98's example: [Collabora](https://apps.obtainium.imranr.dev/redirect?r=obtainium://app/%7B%22id%22%3A%22com.collabora.libreoffice.snapshot%22%2C%22url%22%3A%22https%3A%2F%2Fwww.collaboraoffice.com%2Fdownloads%2FCollabora-Office-Android-Snapshot%2F%22%2C%22author%22%3A%22www.collaboraoffice.com%22%2C%22name%22%3A%22Collabora%20Office%22%2C%22preferredApkIndex%22%3A0%2C%22additionalSettings%22%3A%22%7B%5C%22intermediateLink%5C%22%3A%5B%5D%2C%5C%22customLinkFilterRegex%5C%22%3A%5C%22%5C%22%2C%5C%22filterByLinkText%5C%22%3Afalse%2C%5C%22skipSort%5C%22%3Afalse%2C%5C%22reverseSort%5C%22%3Afalse%2C%5C%22sortByLastLinkSegment%5C%22%3Afalse%2C%5C%22versionExtractWholePage%5C%22%3Afalse%2C%5C%22requestHeader%5C%22%3A%5B%7B%5C%22requestHeader%5C%22%3A%5C%22User-Agent%3A%20Mozilla%2F5.0%20(Linux%3B%20Android%2010%3B%20K)%20AppleWebKit%2F537.36%20(KHTML%2C%20like%20Gecko)%20Chrome%2F114.0.0.0%20Mobile%20Safari%2F537.36%5C%22%7D%5D%2C%5C%22defaultPseudoVersioningMethod%5C%22%3A%5C%22APKLinkHash%5C%22%2C%5C%22trackOnly%5C%22%3Afalse%2C%5C%22versionExtractionRegEx%5C%22%3A%5C%22(%5B0-9%5D%2B%5C%5C%5C%5C.%5B0-9%5D%2B)-snapshot-arm64-v8a-(%5B0-9%5D%2B-%5B0-9%5D%2B-%5B0-9%5D%2B)%5C%22%2C%5C%22matchGroupToUse%5C%22%3A%5C%22%241.%242%5C%22%2C%5C%22versionDetection%5C%22%3Afalse%2C%5C%22useVersionCodeAsOSVersion%5C%22%3Afalse%2C%5C%22apkFilterRegEx%5C%22%3A%5C%22v8a%5C%22%2C%5C%22invertAPKFilter%5C%22%3Afalse%2C%5C%22autoApkFilterByArch%5C%22%3Atrue%2C%5C%22appName%5C%22%3A%5C%22%5C%22%2C%5C%22shizukuPretendToBeGooglePlay%5C%22%3Afalse%2C%5C%22allowInsecure%5C%22%3Afalse%2C%5C%22exemptFromBackgroundUpdates%5C%22%3Afalse%2C%5C%22skipUpdateNotifications%5C%22%3Afalse%2C%5C%22about%5C%22%3A%5C%22%5C%22%2C%5C%22supportFixedAPKURL%5C%22%3Afalse%7D%22%2C%22overrideSource%22%3A%22HTML%22%7D)

Produces version as: `24.04.2024-10-20`

*It's also possible to use other characters between `$1$2`.
E.g.) `$1-$2` produces: `24.04-2024-10-20`
